### ISSUE TITLE
Error when starting up in 5.1 MT environment

### ISSUE THREAD

**Issue by jancalve** [dgradecak/alfresco-mvc#4]
Hi!

We're using version 4.0.0 of alfresco-mvc

When upgrading from Alfresco 4.2.5.4 to 5.1 (enterprise) we encountered a error:
"org.alfresco.error.AlfrescoRuntimeException: 03200119 Keystores are invalid"
...
Caused by: org.springframework.jdbc.BadSqlGrammarException:
### Error querying database.  Cause: org.postgresql.util.PSQLException: ERROR: relation "alf_content_url_encryption" does not exist

Full log is available at https://gist.github.com/jancalve/cec73487f215bec71d81ea9a93a27f82

What happens - from what we can see - is that when the dispatcherWebScript has started up, it emits a refresh event to the Alfresco ContextLoader.configureAndRefreshWebApplicationContext which then attempts to load the keystore and results in the exception.

One of the patches in Alfresco 5.1 is to add a new table - alf_content_url_encryption - which is used when loading the keystore.
I believe this is a timing issue..

In DispatcherWebscript, the WebAplicationContext is initialized via InitializingBean, which happens before any patches are applied to the system and database.

We've manually worked around the problem by executing the Alfresco patch which creates the alf_content_url_encryption table.

Possibly the fix is to initialize the dispatcherWebScript at a later stage so upgrade patches can complete - do you have any ideas?

// JC

**Comment by jancalve** [comment]
I've created a pull request with a successful workaround for this issue. I tested it in an Alfresco 5.1 MT environment successfully.
I'm not sure if there's another more correct way to do this.. Will leave it up to you to decide @dgradecak ;)

**Comment by dgradecak** [comment]
Hi @jancalve,

unfortunately I do not have that enterprise version and cannot test with it. I tested with migrating from 5.0.d to 5.1.e and there is no issue.

Might be easier taht you just disable alfresco-mvc when upgrading. Would that be a possibility? 
I am not keen to change this as there is no easy way for me to test this. If you are aware of a migration path in the community version than I would be happy to test that path. 

If I am not wrong, I think it is better to disable as many modules as you can during upgrade. 
So, if I understand this correctly it is happening only during the upgrade process and once the upgrade is done, alfresco-mvc works correctly? I tested the version 4.0.0 on a clean 5.1 enterprise installation and there was no issue.

**Comment by jancalve** [comment]
Hi @dgradecak ,

The upgrade path we're following is 4.2.3.6 -> 4.2.5.4 (latest service pack) -> 5.1, but for all intents and purposes the problematic path is going from 4.2.5.4 to 5.1 as version 5.0 is where the patch for creating encryption tables is introduced.

Once the upgrade has completed (by manually executing the AlfrescoCreate-ContentUrlEncryptionTables.sql script), alfresco-mvc starts up fine and works correctly.


The problematic patch is called patch.db-V5.0-ContentUrlEncryptionTables which can be found in patch-services-context.xml at https://svn.alfresco.com/repos/alfresco-open-mirror/alfresco/HEAD/root/projects/repository/config/alfresco/patch/patch-services-context.xml

The statements applied can be found in
https://github.com/Alfresco/community-edition/blob/master/projects/repository/config/alfresco/dbscripts/create/org.hibernate.dialect.PostgreSQLDialect/AlfrescoCreate-ContentUrlEncryptionTables.sql

I believe the equivalent upgrade path in the community version would be going from 4.2.f to 5.1.e, as this patch was first introduced in 5.0 of Alfresco - and looking through the sources of 4.2.f there is no mention of this patch.

We can disable alfresco-mvc during the upgrade as a workaround

// JC

**Comment by mauro1855** [comment]
Hi,

First of all thank you for providing this library. We've been using it extensively throughout our project since we started customizing Alfresco.
For information, we've also used your library in community edition 5.1.e and we've had no problems with it, so I guess you can add that version to the compatibility list.

However, now I'm upgrading to community 5.2.g and we've been having some problems. After a few days of digging, I realized the issue seems to be caused by this library. And the issue is exactly the one explained by @jancalve. We have different exceptions, but the library seems to trigger some context event that will trigger some validations on the repo side before it is actually ready to perform those validations.

The first exception I have is with Records Management 2.6.a, where it tries to assert the compatibility of the module by calling repo's (now called "platform") LicenseService that is not yet initialized.

```
Server running at http://localhost:4578
Oct 11, 2017 8:57:28 AM org.apache.catalina.core.StandardContext listenerStart
REPO_ 2017-10-11 08:57:28,953  ERROR [web.context.ContextLoader] [localhost-startStop-1] Context initialization failed
SEVERE: Exception sending context initialized event to listener instance of class org.alfresco.web.app.ContextLoaderListener
java.lang.IllegalStateException: Not bootstrapped
java.lang.IllegalStateException: Not bootstrapped
	at org.alfresco.repo.descriptor.DescriptorServiceImpl.getLicenseDescriptor(DescriptorServiceImpl.java:167)
	at org.alfresco.repo.descriptor.DescriptorServiceImpl.getLicenseDescriptor(DescriptorServiceImpl.java:167)
	at org.alfresco.module.org_alfresco_module_rm.bootstrap.ModuleCompatibilityComponent.onApplicationEvent(ModuleCompatibilityComponent.java:93)
	at org.alfresco.module.org_alfresco_module_rm.bootstrap.ModuleCompatibilityComponent.onApplicationEvent(ModuleCompatibilityComponent.java:93)
	at org.alfresco.module.org_alfresco_module_rm.bootstrap.ModuleCompatibilityComponent.onApplicationEvent(ModuleCompatibilityComponent.java:50)
	at org.alfresco.module.org_alfresco_module_rm.bootstrap.ModuleCompatibilityComponent.onApplicationEvent(ModuleCompatibilityComponent.java:50)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEventInternal(SafeApplicationEventMulticaster.java:214)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEventInternal(SafeApplicationEventMulticaster.java:214)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEvent(SafeApplicationEventMulticaster.java:182)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEvent(SafeApplicationEventMulticaster.java:182)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:334)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:334)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:954)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:954)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:482)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:482)
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:410)
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:410)
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:306)
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:306)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:112)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:112)
	at org.alfresco.web.app.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:70)
	at org.alfresco.web.app.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:70)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4939)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4939)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5434)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5434)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1559)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1559)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1549)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1549)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at java.lang.Thread.run(Thread.java:748)
```
This **always** happens.
I managed to get around this by overriding the LicenseService and force it to initialize when queried. In SDK 3.0.1 this seemed to be enough to make it work.
However I needed to get back to SDK 2.2.0, and there this wasn't enough. In SDK 2.2.0, after I override the LicenseService to prevent that error, I have the following exception: 


```
Server running at http://localhost:4578
REPO_ 2017-10-11 09:13:59,440  ERROR [web.context.ContextLoader] [localhost-startStop-1] Context initialization failed
org.springframework.jdbc.BadSqlGrammarException: 
### Error querying database.  Cause: org.apache.commons.dbcp.SQLNestedException: Borrow prepareStatement from pool failed
### The error may exist in alfresco/ibatis/#resource.dialect#/node-common-SqlMap.xml
### The error may involve alfresco.node.select_StoreRootNodeByRef
### The error occurred while executing a query
### SQL: select             node.id             as id,             node.version        as version,             store.id            as store_id,             store.protocol      as protocol,             store.identifier    as identifier,             node.uuid           as uuid,             node.type_qname_id  as type_qname_id,             node.locale_id      as locale_id,             node.acl_id         as acl_id,             txn.id              as txn_id,             txn.change_txn_id   as txn_change_id,             node.audit_creator  as audit_creator,             node.audit_created  as audit_created,             node.audit_modifier as audit_modifier,             node.audit_modified as audit_modified,             node.audit_accessed  as audit_accessed               from             alf_store store             join alf_node node on (store.root_node_id = node.id)             join alf_transaction txn on (txn.id = node.transaction_id)         where             store.protocol = ? and             store.identifier = ?
### Cause: org.apache.commons.dbcp.SQLNestedException: Borrow prepareStatement from pool failed
; bad SQL grammar []; nested exception is org.apache.commons.dbcp.SQLNestedException: Borrow prepareStatement from pool failed
	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.doTranslate(SQLErrorCodeSQLExceptionTranslator.java:237)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:72)
	at org.mybatis.spring.MyBatisExceptionTranslator.translateExceptionIfPossible(MyBatisExceptionTranslator.java:74)
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:421)
	at com.sun.proxy.$Proxy40.selectOne(Unknown Source)
	at org.mybatis.spring.SqlSessionTemplate.selectOne(SqlSessionTemplate.java:166)
	at org.alfresco.repo.domain.node.ibatis.NodeDAOImpl.selectStoreRootNode(NodeDAOImpl.java:306)
	at org.alfresco.repo.domain.node.AbstractNodeDAOImpl$RootNodesCacheCallbackDAO.findByKey(AbstractNodeDAOImpl.java:896)
	at org.alfresco.repo.domain.node.AbstractNodeDAOImpl$RootNodesCacheCallbackDAO.findByKey(AbstractNodeDAOImpl.java:1)
	at org.alfresco.repo.cache.lookup.EntityLookupCache.getByKey(EntityLookupCache.java:330)
	at org.alfresco.repo.domain.node.AbstractNodeDAOImpl.getRootNode(AbstractNodeDAOImpl.java:759)
	at org.alfresco.repo.node.db.DbNodeServiceImpl.getRootNode_aroundBody14(DbNodeServiceImpl.java:312)
	at org.alfresco.repo.node.db.DbNodeServiceImpl$AjcClosure15.run(DbNodeServiceImpl.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:221)
	at org.alfresco.traitextender.AJExtender.localProceed(AJExtender.java:729)
	at org.alfresco.traitextender.AJProxyTrait.invoke(AJProxyTrait.java:86)
	at com.sun.proxy.$Proxy47.getRootNode(Unknown Source)
	at org.alfresco.repo.virtual.bundle.VirtualNodeServiceExtension.getRootNode(VirtualNodeServiceExtension.java:1211)
	at sun.reflect.GeneratedMethodAccessor231.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.alfresco.traitextender.SingletonExtensionFactory$TraiSingletontHandler.invoke(SingletonExtensionFactory.java:74)
	at com.sun.proxy.$Proxy304.getRootNode(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor232.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.alfresco.traitextender.AJExtender.extendAroundAdvice(AJExtender.java:655)
	at org.alfresco.traitextender.RouteExtensions.ajc$inlineAccessMethod$org_alfresco_traitextender_RouteExtensions$org_alfresco_traitextender_AJExtender$extendAroundAdvice(RouteExtensions.java:1)
	at org.alfresco.traitextender.RouteExtensions.intercept(RouteExtensions.java:85)
	at org.alfresco.repo.node.db.DbNodeServiceImpl.getRootNode(DbNodeServiceImpl.java:310)
	at sun.reflect.GeneratedMethodAccessor231.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at org.alfresco.repo.lock.mem.LockableAspectInterceptor.invoke(LockableAspectInterceptor.java:244)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
	at com.sun.proxy.$Proxy52.getRootNode(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor231.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:317)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at org.alfresco.repo.tenant.MultiTNodeServiceInterceptor.invoke(MultiTNodeServiceInterceptor.java:111)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:204)
	at com.sun.proxy.$Proxy52.getRootNode(Unknown Source)
	at sun.reflect.GeneratedMethodAccessor231.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.alfresco.repo.service.StoreRedirectorProxyFactory$RedirectorInvocationHandler.invoke(StoreRedirectorProxyFactory.java:231)
	at com.sun.proxy.$Proxy79.getRootNode(Unknown Source)
	at org.alfresco.repo.security.authority.AuthorityDAOImpl.getSystemContainer(AuthorityDAOImpl.java:1364)
	at org.alfresco.repo.security.authority.AuthorityDAOImpl.getAuthorityContainer(AuthorityDAOImpl.java:1341)
	at org.alfresco.repo.security.authority.AuthorityDAOImpl.getAuthorityOrNull(AuthorityDAOImpl.java:1321)
	at org.alfresco.repo.security.authority.AuthorityDAOImpl.getAuthorityOrNull(AuthorityDAOImpl.java:1299)
	at org.alfresco.repo.security.authority.AuthorityDAOImpl.authorityExists(AuthorityDAOImpl.java:322)
	at org.alfresco.repo.security.authority.AuthorityServiceImpl.authorityExists(AuthorityServiceImpl.java:638)
	at org.alfresco.module.org_alfresco_module_rm.security.ExtendedSecurityServiceImpl$1$1.execute(ExtendedSecurityServiceImpl.java:154)
	at org.alfresco.module.org_alfresco_module_rm.security.ExtendedSecurityServiceImpl$1$1.execute(ExtendedSecurityServiceImpl.java:150)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:464)
	at org.alfresco.repo.transaction.RetryingTransactionHelper.doInTransaction(RetryingTransactionHelper.java:333)
	at org.alfresco.module.org_alfresco_module_rm.security.ExtendedSecurityServiceImpl$1.doWork(ExtendedSecurityServiceImpl.java:162)
	at org.alfresco.repo.security.authentication.AuthenticationUtil.runAs(AuthenticationUtil.java:555)
	at org.alfresco.module.org_alfresco_module_rm.security.ExtendedSecurityServiceImpl.onApplicationEvent(ExtendedSecurityServiceImpl.java:145)
	at org.alfresco.module.org_alfresco_module_rm.security.ExtendedSecurityServiceImpl.onApplicationEvent(ExtendedSecurityServiceImpl.java:70)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEventInternal(SafeApplicationEventMulticaster.java:214)
	at org.alfresco.repo.management.SafeApplicationEventMulticaster.multicastEvent(SafeApplicationEventMulticaster.java:182)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:334)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:954)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:482)
	at org.springframework.web.context.ContextLoader.configureAndRefreshWebApplicationContext(ContextLoader.java:410)
	at org.springframework.web.context.ContextLoader.initWebApplicationContext(ContextLoader.java:306)
	at org.springframework.web.context.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:112)
	at org.alfresco.web.app.ContextLoaderListener.contextInitialized(ContextLoaderListener.java:70)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4939)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5434)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1559)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1549)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.commons.dbcp.SQLNestedException: Borrow prepareStatement from pool failed
	at org.apache.commons.dbcp.PoolingConnection.prepareStatement(PoolingConnection.java:113)
	at org.apache.commons.dbcp.DelegatingConnection.prepareStatement(DelegatingConnection.java:281)
	at org.apache.commons.dbcp.PoolingDataSource$PoolGuardConnectionWrapper.prepareStatement(PoolingDataSource.java:313)
	at sun.reflect.GeneratedMethodAccessor237.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.hibernate.jdbc.BorrowedConnectionProxy.invoke(BorrowedConnectionProxy.java:150)
	at com.sun.proxy.$Proxy38.prepareStatement(Unknown Source)
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.instantiateStatement(PreparedStatementHandler.java:79)
	at org.apache.ibatis.executor.statement.BaseStatementHandler.prepare(BaseStatementHandler.java:88)
	at org.apache.ibatis.executor.statement.RoutingStatementHandler.prepare(RoutingStatementHandler.java:58)
	at org.apache.ibatis.executor.SimpleExecutor.prepareStatement(SimpleExecutor.java:76)
	at org.apache.ibatis.executor.SimpleExecutor.doQuery(SimpleExecutor.java:61)
	at org.apache.ibatis.executor.BaseExecutor.queryFromDatabase(BaseExecutor.java:303)
	at org.apache.ibatis.executor.BaseExecutor.query(BaseExecutor.java:154)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:102)
	at org.apache.ibatis.executor.CachingExecutor.query(CachingExecutor.java:82)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:120)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectList(DefaultSqlSession.java:113)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.selectOne(DefaultSqlSession.java:73)
	at sun.reflect.GeneratedMethodAccessor233.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:408)
	... 83 more
Caused by: org.h2.jdbc.JdbcSQLException: Table "ALF_STORE" not found; SQL statement:
select
            node.id             as id,
            node.version        as version,
            store.id            as store_id,
            store.protocol      as protocol,
            store.identifier    as identifier,
            node.uuid           as uuid,
            node.type_qname_id  as type_qname_id,
            node.locale_id      as locale_id,
            node.acl_id         as acl_id,
            txn.id              as txn_id,
            txn.change_txn_id   as txn_change_id,
            node.audit_creator  as audit_creator,
            node.audit_created  as audit_created,
            node.audit_modifier as audit_modifier,
            node.audit_modified as audit_modified,
            node.audit_accessed  as audit_accessed
     
        from
            alf_store store
            join alf_node node on (store.root_node_id = node.id)
            join alf_transaction txn on (txn.id = node.transaction_id)
        where
            store.protocol = ? and
            store.identifier = ? [42102-190]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.command.Parser.readTableOrView(Parser.java:5293)
	at org.h2.command.Parser.readTableFilter(Parser.java:1234)
	at org.h2.command.Parser.parseSelectSimpleFromPart(Parser.java:1873)
	at org.h2.command.Parser.parseSelectSimple(Parser.java:1982)
	at org.h2.command.Parser.parseSelectSub(Parser.java:1867)
	at org.h2.command.Parser.parseSelectUnion(Parser.java:1688)
	at org.h2.command.Parser.parseSelect(Parser.java:1676)
	at org.h2.command.Parser.parsePrepared(Parser.java:432)
	at org.h2.command.Parser.parse(Parser.java:304)
	at org.h2.command.Parser.parse(Parser.java:276)
	at org.h2.command.Parser.prepareCommand(Parser.java:241)
	at org.h2.engine.Session.prepareLocal(Session.java:460)
	at org.h2.engine.Session.prepareCommand(Session.java:402)
	at org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1188)
	at org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:72)
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:276)
	at org.apache.commons.dbcp.PoolingConnection.makeObject(PoolingConnection.java:285)
	at org.apache.commons.pool.impl.GenericKeyedObjectPool.borrowObject(GenericKeyedObjectPool.java:1220)
	at org.apache.commons.dbcp.PoolingConnection.prepareStatement(PoolingConnection.java:107)
	... 106 more

```

This only happens if I purge my system, namely if I delete the database. Normally alfresco would run the scripts to initialize it, but here the event is triggered and repo tries to access the database before it is actually initialized/ready.

If I have the database already initialized, than it works OK.

The bottomline is that the library really is triggering repo to perform actions before it has been completely initialized, so I'll try to take a look at @jancalve pull request and also propose a solution if possible.

Thanks,
Mauro

**Comment by dgradecak** [comment]
Hi Mauro,

are you also in a multi tenant environment? Did you try the same upgrade path without RM? I will try this upgrade path in the next days without RM and if the issue happens I will do my best to fix it.

Btw, are you on v4 of alfresco mvc? The version 5 is still a snapshot but it is being used in production and simplifies a lot the configuration as it leverages java config.

Please let me know about your progress and I will come back on this issue quickly too,

**Comment by mauro1855** [comment]
Hi,

No, our environment is single tenant, no MT configuration is done. I'm using the v4 of mvc, yes.
Sorry, I didn't try without RM, because my code requires it.

The good news is that after looking at @jancalve code, **I manage to solve the issue and I will be doing a pull request tomorrow**. The premise is the same as @jancalve, I bind something to the repositoryEndBootstrapBean so that MVC only fires when repo finishes bootstrapping. However my approach is a bit different than his. However in the end both our approaches do the same, I don't know why his didn't work for him.

Even if you don't accept the pull request (since you are working on a new branch), I'll still make it so you see how I solved it.

**Comment by dgradecak** [comment]
Hi Mauro, 

thank you for helping. I will check correctly the pull request you did but before I will need to spend some time to reproduce the issue first. After that I want to check it with the alfresco-mvc v5.0.0 as the usage is completely different and there is no more module at bootstrap actually. So I wonder it that new mechanism resolves the same issue.

Thank you once more and I will get back to this very quickly.

**Comment by lorelj2** [comment]
Hi @mauro1855. 

thank you again for helping. The issue is with initializing the DispatcherWebscript WebApplicationContext. At the end of initializing WebApplicationContext it triggers refresh event. Many of Alfresco components have context refresh event listener, but they are checking where is that refresh event coming from, so on clean Alfresco without additional modules there are no problems. The problem is that Alfresco Record management module also has context refresh listener, but it does not check from which context does that refresh event comes from (we presume many other components are also doing it like this) so it got triggered by child WebApplicationContext while bootstrapping process hasn't finished yet.

To fix the problem we moved the initialization of DispatcherWebscript and its WebApplicationContext to application context refresh event that gets triggered after bootstrapping process. The fix is applied to branch hotfix/4.0.1. Could you please test it with hotfix?

Also, we are planning to release Alfresco MVC 5 soon. Are you interested in migration to a newer version?

**Comment by mauro1855** [comment]
Hi @lorelj2,

Your explanation makes total sense considering what we saw in the logs.
I also did my own fix that you can see on the pull request I made (the pull request is actually huge, I guess my repository had more modifications made by my team mates, but what I changed was actually just a couple of classes, I commented this in the PR). I can see if I can find the time to check your hotfix.

Regarding the MVC 5, I guess when it is release we could try.
